server/src/main/kotlin/suwayomi/tachidesk/manga/impl/ChapterFile.kt

```diff
@@ -0,0 +1,85 @@
+import suwayomi.tachidesk.manga.impl.download.model.DownloadChapter
+import suwayomi.tachidesk.manga.impl.util.getChapterDir
+import java.io.File
+import java.io.FileOutputStream
+import java.io.InputStream
+import java.util.zip.ZipEntry
+import java.util.zip.ZipFile
+import java.util.zip.ZipInputStream
+import java.util.zip.ZipOutputStream
+
+object ChapterFile {
+
+    fun isFile(mangaId: Int, chapterId: Int): Boolean {
+        val chapterDir = getChapterDir(mangaId, chapterId)
+        return File("$chapterDir.cbz").exists()
+    }
+
+    fun fetchImage(mangaId: Int, chapterId: Int, index: Int): Pair<InputStream, String> {
+        val chapterPath = getChapterDir(mangaId, chapterId)
+        val cbzPath = "$chapterPath.cbz"
+        val zipFile = ZipFile(cbzPath)
+        try {
+            val zipEntry = zipFile.entries().toList().sortedWith(compareBy({ it.name }, { it.name }))[index]
+            val inputStream = zipFile.getInputStream(zipEntry)
+            return Pair(inputStream, zipEntry.name)
+        } catch (ex: ArrayIndexOutOfBoundsException) {
+            // Panic! extract everything, delete outdated cbzFile so that download retry works properly
+            zipFile.close()
+            extractFromCbz(chapterPath, cbzPath)
+            File(cbzPath).delete()
+            File(chapterPath).delete()
+            throw ArrayIndexOutOfBoundsException()
+        }
+    }
+
+    private fun extractFromCbz(chapterDir: String, cbzPath: String) {
+        val chapterFolder = File(chapterDir)
+        if (!chapterFolder.exists()) chapterFolder.mkdir()
+        val zipInputStream = ZipInputStream(File(cbzPath).inputStream())
+        var zipEntry = zipInputStream.nextEntry
+
+        while (zipEntry != null) {
+            val file = File(chapterFolder, zipEntry.name)
+            if (!file.exists()) {
+                file.parentFile.mkdirs()
+                file.createNewFile()
+            }
+
+            val outputStream = FileOutputStream(file)
+            zipInputStream.copyTo(outputStream)
+            outputStream.close()
+            zipEntry = zipInputStream.nextEntry
+        }
+    }
+
+    fun convertToCbz(download: DownloadChapter?) {
+        if (download?.mangaId != null) {
+            val chapterDir = getChapterDir(download.mangaId, download.chapter.id)
+            val outputFile = File("$chapterDir.cbz")
+
+            if (!outputFile.exists()) outputFile.createNewFile()
+
+            ZipOutputStream(outputFile.outputStream()).use { zipOut ->
+                transaction {
+                    val folder = File(chapterDir)
+                    if (folder.isDirectory) {
+                        val files = folder.listFiles()?.sortedBy { it.name }?.forEach {
+                            val entry = ZipEntry(it.name)
+                            zipOut.putNextEntry(entry)
+                            it.inputStream().copyTo(zipOut)
+                            zipOut.closeEntry()
```

Comment:
Make sure this gets closed, with try {} finally {}

Reply:
or with kotling use syntax preferebly

Reply:
What should I specifically do if closeEntry fails? Propagate the error so the downloader dies? or do something else?

Reply:
Propagate the error, don't try to catch it, just put .closeEntry in the finally block, try catch finally doesn't require a catch.

Reply:
Did this under the latest comment, sorry missed this comment. I am marking this as resolved